### PR TITLE
ecs_task_observer: initial structure

### LINKED ISSUE #1395: [Feature Request] [Extension] ECS Service Discovery for Prometheus
**Is your feature request related to a problem? Please describe.**
Customers might want to perform ECS Service Discovery for Prometheus metric scraping in their pipeline. Currently, there is no option to perform this functionality. 

**Describe the solution you'd like**
Our proposal is to create a new extension that performs this ECS Service Discovery logic for the Prometheus receiver. This extension would invoke a separate process that would query ECS API for task metadata and output the discovered list of scrape targets to a defined output file that the Prometheus receiver can scrape from (configured via `file_sd_configs`). This functionality would be similar to the ECS Service Discovery logic in AWS CloudWatch. More info of how this is done in the CloudWatch Agent can be found in [this README](https://github.com/aws/amazon-cloudwatch-agent/blob/master/internal/ecsservicediscovery/README.md) and the [online documentation](https://docs.aws.amazon.com/en_en/AmazonCloudWatch/latest/monitoring/ContainerInsights-Prometheus-Setup-autodiscovery-ecs.html).

**Describe alternatives you've considered**
- Implement as part of Prometheus receiver to support ECS service discovery. There was an earlier proposal for this that was rejected because the ECS API can be throttled and is expensive to call.
- Implement this service discovery in a wrapper around Otel Collector. This would not make the functionality available to all customers using the collector.

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
**Description:**
Adding a feature - These changes introduce a new receiver creator-compatible watch observer for ECS tasks.  It differs from the existing ECS observer in that it supports only sidecar deployments and doesn't have a hard dependency on the Prometheus receiver, instead being able to sync with a receiver creator instance by listing [container endpoints](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/957eb18037f000f5740c4044fdf5ffeee4311ebc/extension/observer/endpoints.go#L166).  This approach was landed on after reviewing the existing and ECS observer config and implementation and from following the guidance of the initial observer's author: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5316#issuecomment-925545114.  It does not resolve [the outstanding item](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/1395#issuecomment-747715986) for #1395, as I think observer endpoint creation should be added to the existing observer as well but should be done in unrelated changes.

**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/5316

**Testing:**
~No tests as this time since the changes just include the initial types and readme.~ Basic factory test to help ensure receiver creator type compatibility.*

**Documentation:**
Adding a readme with high level plan and config documentation generated by project configschema tool.

cc @alolita

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@Aneurysm9 would be good someone from AWS to review/own(coown) this

File: extension/observer/ecstaskobserver/extension.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+var _ observer.Observable = (*ecsTaskObserver)(nil)
+
+type ecsTaskObserver struct {
+	logger *zap.Logger
+	config *Config
+}
+
+func (e *ecsTaskObserver) Start(_ context.Context, host component.Host) error {
+	return nil
+}
```
Comment (bogdandrutu) [inline extension/observer/ecstaskobserver/extension.go:36]: use https://pkg.go.dev/go.opentelemetry.io/collector@v0.38.0/component/componenthelper to build the extension so no need to implement funcs like Start/Shutdown if not needed.

File: extension/observer/ecstaskobserver/factory.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+func createExtension(
+	_ context.Context,
+	params component.ExtensionCreateSettings,
+	cfg config.Extension,
+) (component.Extension, error) {
+	obsCfg := cfg.(*Config)
+	e := &ecsTaskObserver{
+		config: obsCfg,
+		logger: params.Logger,
```
Comment (bogdandrutu) [inline extension/observer/ecstaskobserver/factory.go:50]: We recommend to pass the entire "Telemetry".

**Review by bogdandrutu — APPROVED** [review]

**Review by Aneurysm9 — APPROVED** [review]
This looks like a nice addition.  @rakyll I know you've talked about ways to improve ECS task detection, does this align with the direction you'd expect?

**Comment by Aneurysm9** [comment]
> Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml
verifyAllModulesInSet failed: Module github.com/open-telemetry/opentelemetry-collector-contrib/extension/observer/ecstaskobserver (defined in /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/extension/observer/ecstaskobserver/go.mod) is not listed in any module set.

Please add this new module to `versions.yaml`.

File: extension/observer/ecstaskobserver/factory.go (hunk 2 above)
Comment (rmfitzpatrick) [reply extension/observer/ecstaskobserver/factory.go]: updated

File: extension/observer/ecstaskobserver/extension.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+var _ component.Extension = (*ecsTaskObserver)(nil)
+var _ observer.Observable = (*ecsTaskObserver)(nil)
+
+type ecsTaskObserver struct {
+	component.Extension
+	config    *Config
+	telemetry component.TelemetrySettings
+}
+
+func (e *ecsTaskObserver) ListAndWatch(notify observer.Notify) {}
```
Comment (MovieStoreGuy) [inline extension/observer/ecstaskobserver/extension.go:32]: Can ask you to comment your exported methods.

File: extension/observer/ecstaskobserver/factory_test.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+func TestFactoryCreatedExtensionIsObservable(t *testing.T) {
+	etoFactory := NewFactory()
+	eto, err := etoFactory.CreateExtension(
+		context.Background(), componenttest.NewNopExtensionCreateSettings(), etoFactory.CreateDefaultConfig(),
+	)
+	require.NoError(t, err)
+	require.NotNil(t, eto)
+	asObservable, ok := eto.(observer.Observable)
```
Comment (MovieStoreGuy) [inline extension/observer/ecstaskobserver/factory_test.go:34]: I suggest using https://pkg.go.dev/github.com/stretchr/testify/assert#Implements to check if it implements the interface.

However, this feels like a redundant tests considering there is a compile time check

**Review by MovieStoreGuy — APPROVED** [review]

File: extension/observer/ecstaskobserver/extension.go (hunk 3 above)
Comment (rmfitzpatrick) [reply extension/observer/ecstaskobserver/extension.go]: Updated and made a revision for the actual desired helper type implementation.

File: extension/observer/ecstaskobserver/factory_test.go (hunk 4 above)
Comment (rmfitzpatrick) [reply extension/observer/ecstaskobserver/factory_test.go]: This test is for the factory method which wouldn't enforce the returned helper type at compile time, though the extension itself does have an added implementation check that could make this redundant.  Given the level of indirection I think it's a helpful assurance.

Updated to use the implementation helper.

**Comment by tigrannajaryan** [comment]
@rmfitzpatrick please populate the commit message in future PRs (the PR description is typically great to use the commit message). I copy/pates this time before merging.

**Pull request merged**